analyzers/src/SonarAnalyzer.Common/Helpers/MethodParameterLookupBase.cs

```diff
@@ -18,115 +18,135 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
-namespace SonarAnalyzer.Helpers
+namespace SonarAnalyzer.Helpers;
+
+public interface IMethodParameterLookup
 {
-    public interface IMethodParameterLookup
+    bool TryGetSymbol(SyntaxNode argument, out IParameterSymbol parameter);
+    bool TryGetSyntax(IParameterSymbol parameter, out ImmutableArray<SyntaxNode> expressions);
+    bool TryGetSyntax(string parameterName, out ImmutableArray<SyntaxNode> expressions);
+    bool TryGetNonParamsSyntax(IParameterSymbol parameter, out SyntaxNode expression);
+}
+
+// This should come from the Roslyn API (https://github.com/dotnet/roslyn/issues/9)
+internal abstract class MethodParameterLookupBase<TArgumentSyntax> : IMethodParameterLookup
+    where TArgumentSyntax : SyntaxNode
+{
+    private readonly SeparatedSyntaxList<TArgumentSyntax>? argumentList;
+
+    protected abstract SyntaxToken? GetNameColonArgumentIdentifier(TArgumentSyntax argument);
+    protected abstract SyntaxNode Expression(TArgumentSyntax argument);
+
+    public IMethodSymbol MethodSymbol { get; }
+    private ImmutableArray<IMethodSymbol> MethodSymbolOrCandidates { get; }
+
+    protected MethodParameterLookupBase(SeparatedSyntaxList<TArgumentSyntax>? argumentList, SymbolInfo? methodSymbolInfo)
+     : this(argumentList, methodSymbolInfo?.Symbol as IMethodSymbol, methodSymbolInfo?.AllSymbols().OfType<IMethodSymbol>()) { }
```

Comment:
The constructor is not really sound. If either argumentList or methodSymbolInfo is not known, the whole parameter lookup doesn't make sense:
```suggestion
    protected MethodParameterLookupBase(SeparatedSyntaxList<TArgumentSyntax> argumentList, SymbolInfo methodSymbolInfo)
     : this(argumentList, methodSymbolInfo.Symbol as IMethodSymbol, methodSymbolInfo.AllSymbols().OfType<IMethodSymbol>()) { }
```
We might also do this change later as it will probably affect a lot of code.

Reply:
Indentation is off.

Reply:
I don't think we can make this change without breaking a bunch of existing code and tests. So yes, let's do it at a later time.